dao/src/main/java/com/epam/esm/dao/impl/CertificateDaoImpl.java

```diff
@@ -0,0 +1,103 @@
+package com.epam.esm.dao.impl;
+
+import com.epam.esm.dao.CertificateDao;
+import com.epam.esm.dao.exception.WrongInsertDataException;
+import com.epam.esm.dao.mapper.ColumnName;
+import com.epam.esm.domain.entity.Certificate;
+import com.epam.esm.dao.mapper.CertificateMapper;
+import org.springframework.jdbc.core.JdbcTemplate;
+import org.springframework.jdbc.support.GeneratedKeyHolder;
+import org.springframework.jdbc.support.KeyHolder;
+import org.springframework.stereotype.Component;
+
+import javax.sql.DataSource;
+import java.sql.PreparedStatement;
+import java.util.List;
+import java.util.Objects;
+import java.util.Optional;
+
+@Component
+public class CertificateDaoImpl implements CertificateDao {
+
+    private static final String UPDATE_CERTIFICATE = "UPDATE gift_certificate  SET " +
+            "name = COALESCE(?, name), description = COALESCE(?, description), " +
+            "price = COALESCE(?, price), duration = COALESCE(?, duration) WHERE (`id` = ?)";
+    private static final String INSERT_INTO_CERT_HAS_TAG = "INSERT INTO gift_certificate_has_tag " +
+            "(`gift_certificate_id`, `tag_id`) VALUES (?, ?)";
+    private static final String SELECT_BY_SEARCH_PARAMETERS = "SELECT c.id, c.name, c.description, c.price, " +
+            "c.duration, c.create_date, c.last_update_date " +
+            "FROM gift_certificate AS c " +
+            "JOIN gift_certificate_has_tag AS cht ON c.id = cht.gift_certificate_id " +
+            "JOIN tag AS t on cht.tag_id=t.id ";
+    private static final String INSERT_CERTIFICATE = "INSERT INTO gift_certificate " +
+            "(`name`, `description`, `price`, `duration`) VALUES (?, ?, ?, ?)";
+    private static final String SELECT_CERTIFICATE_BY_ID = "SELECT c.id, c.name, c.description, c.price, " +
+            "c.duration, c.create_date, c.last_update_date FROM gift_certificate AS c WHERE id=?";
+    private static final String SELECT_ALL_CERTIFICATES = "SELECT c.id, c.name, c.description, c.price, " +
+            "c.duration, c.create_date, c.last_update_date FROM gift_certificate AS c";
+    private static final String DELETE_CERTIFICATE = "DELETE FROM gift_certificate WHERE id = ?";
+
+    private final JdbcTemplate jdbcTemplate;
+    private final CertificateMapper certificateMapper;
+
+    public CertificateDaoImpl(CertificateMapper certificateMapper, DataSource dataSource) {
+        this.jdbcTemplate = new JdbcTemplate(dataSource);

```

Comment:
Why this class create its dependencies itself instead of DI? I believe this change caused by one of the comments. But take a look at declaration for PlatformTransactionManager in your config file for example.

Reply:
yes, you are right.  thank you. fixed